test/common/secret/sds_api_test.cc

```diff
@@ -115,6 +115,50 @@ TEST_F(SdsApiTest, DynamicTlsCertificateUpdateSuccess) {
   handle->remove();
 }
 
+// Tests SDS's use of the delta variant of onConfigUpdate().
+TEST_F(SdsApiTest, DeltaUpdateSuccess) {
+  NiceMock<Server::MockInstance> server;
+  NiceMock<Init::MockManager> init_manager;
+  envoy::api::v2::core::ConfigSource config_source;
+  TlsCertificateSdsApi sds_api(
+      server.localInfo(), server.dispatcher(), server.random(), server.stats(),
+      server.clusterManager(), init_manager, config_source, "abc.com", []() {}, *api_);
+
+  NiceMock<Secret::MockSecretCallbacks> secret_callback;
+  auto handle =
+      sds_api.addUpdateCallback([&secret_callback]() { secret_callback.onAddOrUpdateSecret(); });
+
+  std::string yaml =
+      R"EOF(
+  name: "abc.com"
+  tls_certificate:
+    certificate_chain:
+      filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/selfsigned_cert.pem"
+    private_key:
+      filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/selfsigned_key.pem"
+    )EOF";
+  envoy::api::v2::auth::Secret typed_secret;
+  MessageUtil::loadFromYaml(TestEnvironment::substitute(yaml), typed_secret);
+  Protobuf::RepeatedPtrField<envoy::api::v2::Resource> secret_resources;
+  secret_resources.Add()->mutable_resource()->PackFrom(typed_secret);
+
+  EXPECT_CALL(secret_callback, onAddOrUpdateSecret());
+  sds_api.onConfigUpdate(secret_resources, {}, "");
+
+  Ssl::TlsCertificateConfigImpl tls_config(*sds_api.secret(), *api_);
+  const std::string cert_pem =
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/selfsigned_cert.pem";
+  EXPECT_EQ(TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(cert_pem)),
+            tls_config.certificateChain());
+
+  const std::string key_pem =
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/selfsigned_key.pem";
+  EXPECT_EQ(TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(key_pem)),
+            tls_config.privateKey());
+
```

Comment:
Does there need to be some assertions here, additional `onConfigUpdate()`, etc?

Reply:
I don't really know. This is why I left these tests out originally; I don't know anything about what SDS does, so all I can do is copy the existing tests.

Reply:
Do we have a test covering the new behavior? We probably need at least some basic functional coverage.

Reply:
@fredlas this is the main thing blocking merging right now, WDYT?

Reply:
There is no new behavior for SDS; the "delta" version of its onConfigUpdate is a direct passthrough to the SotW version. I copied this test just to get coverage of those couple of passthrough lines. Given how relatively complex the test is, maybe it's not worth keeping it around just for coverage of a couple of trivial lines?

Reply:
Is there a way to just unit test this single method, the one doing the pass through?

Reply:
It's passing through to another method of the same class, so it can't be mocked away. So, no.

Reply:
Hmm, aren't they both virtual methods? That should be possible to mock.

Reply:
I don't understand how that would work. There is a class `C`, a function `C::f(a)` and a function
```
C::f(a,b) {
  f(a)
}
```
In the test, there is instance `C c`, which you are calling `c.f(a,b)` on. How can that lead to something other than the real `C::f(a)` getting called?

Reply:
Create
```
class D : public C {
  MOCK_METHOD1(f, void(int));
}
```
Now, you create an instance of D, invoke `d.f(a,b)` and hook your mock with the above. Could something like that work?

Reply:
Done.